sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala

```diff
@@ -235,12 +236,13 @@ case class ExpressionEncoder[T](
 
     val plan = Project(Alias(unbound, "")() :: Nil, LocalRelation(schema))
     val analyzedPlan = SimpleAnalyzer.execute(plan)
+    val optimizedPlan = SimplifyCasts(analyzedPlan)
```

Comment:
Probably not a `SQLContext` but a `CatalystConfig` would be reasonable.  I wonder if it should be a different setting than SQL case sensitivity resolution?

On one hand, Scala/Java are always case sensitive so it seems reasonable to preserve that.  On the other hand if you loading from something like hive it would be annoying to have to fix all the columns by hand.

@rxin, thoughts?